File: harness/ts-oracle/oracle.mjs
/**
 * Change-impact oracle for a TypeScript method signature change.
 *
 * The TS sibling of java-oracle/Oracle.java: given Type#method, emits every
 * site that must change — the declaration, every implementation/override
 * (compiler-resolved via the language service, NOT name matching), and every
 * call site that binds to a family member, attributed to its enclosing
 * method/function. Ground truth is production sources only (src/).
 *
 * Usage:
 *   node oracle.mjs --project <tsconfig.json> --repo <root> \
 *        --target 'QueryRunner#addColumn' --out out.json
 *   node oracle.mjs --project <tsconfig.json> --repo <root> --index idx.json
 *     (line->name index for rescore_java.py, same JSON shape as the Spoon one)
 */
import { Project, Node } from "ts-morph";
import { writeFileSync } from "node:fs";
import path from "node:path";

const args = {};
for (let i = 2; i < process.argv.length; i++) {
    if (process.argv[i].startsWith("--")) args[process.argv[i].slice(2)] = process.argv[++i];
}
const need = (k) => { if (!args[k]) { console.error(`missing --${k}`); process.exit(1); } return args[k]; };

const repo = path.resolve(need("repo"));
const project = new Project({ tsConfigFilePath: need("project") });
const rel = (fp) => path.relative(repo, fp);
const inSrc = (fp) => {
    const r = rel(fp);
    return !r.startsWith("..") && !r.startsWith("node_modules") &&
        !r.includes("/test/") && !r.startsWith("test/") &&
        !/\.(spec|test)\.ts$/.test(r);
};

/** Name of the enclosing method/function/class-member for a node. */
function enclosingSite(node) {
    let cur = node;
    while (cur) {
        if (Node.isMethodDeclaration(cur) || Node.isMethodSignature(cur) ||
            Node.isFunctionDeclaration(cur) || Node.isGetAccessorDeclaration(cur) ||
            Node.isSetAccessorDeclaration(cur)) {
            const n = cur.getName();
            if (n) return n;
        }
        if (Node.isConstructorDeclaration(cur)) {
            const cls = cur.getParent();
            return (Node.isClassDeclaration(cls) && cls.getName()) || "constructor";
        }
        // Class fields (arrow-function members) are terminal — the method-like
        // unit. Variable declarations inside bodies are NOT: keep climbing to
        // the enclosing method, matching the Java oracle's attribution.
        if (Node.isPropertyDeclaration(cur)) {
            const n = cur.getName();
            if (n) return n;
        }
        cur = cur.getParent();
    }
    return "<module>";
}

// ---- index mode: relpath -> [[startLine, endLine, name], ...] --------------
if (args.index) {
    const idx = {};
    for (const sf of project.getSourceFiles()) {
        if (!inSrc(sf.getFilePath())) continue;
        const spans = [];
        sf.forEachDescendant((n) => {
            if (Node.isMethodDeclaration(n) || Node.isFunctionDeclaration(n) ||
                Node.isConstructorDeclaration(n) ||
                Node.isGetAccessorDeclaration(n) || Node.isSetAccessorDeclaration(n)) {
                const name = Node.isConstructorDeclaration(n)
                    ? (n.getParent()?.getName?.() ?? "constructor")
                    : n.getName();
                if (name) spans.push([n.getStartLineNumber(), n.getEndLineNumber(), name]);
            }
        });
        if (spans.length) idx[rel(sf.getFilePath())] = spans;
    }
    writeFileSync(args.index, JSON.stringify(idx));
    console.log(`[index] files=${Object.keys(idx).length} -> ${args.index}`);
    process.exit(0);
}

// ---- change-impact mode -----------------------------------------------------
const target = need("target");
const [typeName, methodName] = target.split("#");
if (!methodName) { console.error("--target must be Type#method"); process.exit(1); }

// 1. Declaration: the named interface/class member.
let decl = null;
for (const sf of project.getSourceFiles()) {
    if (!inSrc(sf.getFilePath())) continue;
    for (const t of [...sf.getInterfaces(), ...sf.getClasses()]) {
        if (t.getName() !== typeName) continue;
        const m = t.getMember?.(methodName) ?? t.getMethod?.(methodName);
        if (m) decl = m;
    }
}
if (!decl) { console.error(`target ${target} not found in project sources`); process.exit(1); }

const ls = project.getLanguageService();
const familySites = new Set();
const familyNodes = [decl];
familySites.add(`${rel(decl.getSourceFile().getFilePath())}:${methodName}`);

// 2. Family: compiler-resolved implementations of the member.
for (const impl of ls.getImplementations(decl.getNameNode() ?? decl)) {
    const node = impl.getNode();
    const fp = impl.getSourceFile().getFilePath();
    if (!inSrc(fp)) continue;
    familySites.add(`${rel(fp)}:${methodName}`);
    familyNodes.push(node);
}

// 3. Callers: references to the declaration or any family member that occur
// in call position, attributed to the enclosing method/function.
const callerSites = new Set();
const seenRefPos = new Set();
for (const fam of familyNodes) {
    const nameNode = fam.getNameNode?.() ?? fam;
    let refs;
    try { refs = ls.findReferencesAsNodes(nameNode); } catch { continue; }
    for (const ref of refs) {
        const fp = ref.getSourceFile().getFilePath();
        if (!inSrc(fp)) continue;
        const key = fp + ":" + ref.getStart();
        if (seenRefPos.has(key)) continue;
        seenRefPos.add(key);
        // Call position: obj.method(...) or method(...)
        const parent = ref.getParent();
        let isCall = false;
        if (Node.isPropertyAccessExpression(parent)) {
            const gp = parent.getParent();
            isCall = Node.isCallExpression(gp) && gp.getExpression() === parent;
        } else if (Node.isCallExpression(parent)) {
            isCall = parent.getExpression() === ref;
        }
        if (!isCall) continue;
        const site = `${rel(fp)}:${enclosingSite(ref)}`;
        // A family member calling a sibling is still a caller site only if the
        // enclosing method is not itself in the family for the same file:name.
        if (!familySites.has(site)) callerSites.add(site);
    }
}

const gt = [...new Set([...familySites, ...callerSites])].sort();
const out = {
    target,
    family: [...familySites].sort(),
    callers: [...callerSites].sort(),
    ground_truth: gt,
    stats: { family: familySites.size, resolved_call_hits: callerSites.size, gt_sites: gt.length },
};
writeFileSync(need("out"), JSON.stringify(out, null, 2) + "\n");
console.log(JSON.stringify(out.stats));

File: harness/py-oracle/oracle.py
"""Change-impact oracle for a Python method signature change (Jedi-resolved).

The Python sibling of java-oracle/Oracle.java and ts-oracle/oracle.mjs: given
Type.method, emit every site that must change if that method's signature
changes --

  1. the declaration + every override in the subtype closure (the family), and
  2. every call site that Jedi resolves to a family member, attributed to its
     enclosing function/method.

Family closure is AST-based (transitive subclasses defining the method).
Caller resolution is Jedi's (type inference, project-wide), NOT name matching:
a call to an unrelated same-named method on a different type is not included --
Jedi resolves it elsewhere and it is dropped. Ground truth is production
sources only (the package dir passed as --src, tests excluded).

Site format is "<repo-relative-path>:<enclosing-def-name>" (harness Site form).

Usage:
  python oracle.py --src <pkgdir> --repo <root> --target 'Console.print' --out out.json
  python oracle.py --src <pkgdir> --repo <root> --index idx.json   # line->name index
"""
from __future__ import annotations

import argparse
import ast
import json
from pathlib import Path

import jedi


def py_files(src: Path) -> list[Path]:
    out = []
    for p in src.rglob("*.py"):
        rel = p.as_posix()
        if "/test" in rel or "/tests/" in rel or p.name.startswith("test_"):
            continue
        out.append(p)
    return out


def enclosing_def(tree: ast.Module, line: int) -> str:
    """Name of the innermost def/class enclosing a 1-based line."""
    best_name = "<module>"
    best_span = 10**9
    for node in ast.walk(tree):
        if isinstance(node, (ast.FunctionDef, ast.AsyncFunctionDef)):
            end = getattr(node, "end_lineno", node.lineno)
            if node.lineno <= line <= end:
                span = end - node.lineno
                if span < best_span:
                    best_span, best_name = span, node.name
    return best_name


def build_class_index(files: list[Path]) -> tuple[dict, dict]:
    """Return (class_defs, subclasses).

    class_defs: simple class name -> list of (file, ast.ClassDef).
    subclasses: simple base name -> set of simple subclass names (direct).
    """
    class_defs: dict[str, list] = {}
    subclasses: dict[str, set] = {}
    for f in files:
        try:
            tree = ast.parse(f.read_text(), filename=str(f))
        except SyntaxError:
            continue
        for node in ast.walk(tree):
            if isinstance(node, ast.ClassDef):
                class_defs.setdefault(node.name, []).append((f, node))
                for base in node.bases:
                    bname = base.id if isinstance(base, ast.Name) else (
                        base.attr if isinstance(base, ast.Attribute) else None)
                    if bname:
                        subclasses.setdefault(bname, set()).add(node.name)
    return class_defs, subclasses


def method_in_class(cls: ast.ClassDef, method: str) -> ast.stmt | None:
    for item in cls.body:
        if isinstance(item, (ast.FunctionDef, ast.AsyncFunctionDef)) and item.name == method:
            return item
    return None


def transitive_subclasses(root: str, subclasses: dict) -> set[str]:
    seen, frontier = set(), [root]
    while frontier:
        cur = frontier.pop()
        for sub in subclasses.get(cur, ()):
            if sub not in seen:
                seen.add(sub)
                frontier.append(sub)
    return seen


def index_mode(src: Path, repo: Path, out: Path) -> None:
    idx: dict[str, list] = {}
    for f in py_files(src):
        try:
            tree = ast.parse(f.read_text(), filename=str(f))
        except SyntaxError:
            continue
        spans = []
        for node in ast.walk(tree):
            if isinstance(node, (ast.FunctionDef, ast.AsyncFunctionDef)):
                end = getattr(node, "end_lineno", node.lineno)
                spans.append([node.lineno, end, node.name])
        if spans:
            idx[f.relative_to(repo).as_posix()] = spans
    out.write_text(json.dumps(idx))
    print(f"[index] files={len(idx)} -> {out}")


def main() -> None:
    ap = argparse.ArgumentParser()
    ap.add_argument("--src", required=True, help="package dir to scan")
    ap.add_argument("--repo", required=True, help="corpus root (for relpaths)")
    ap.add_argument("--target", help="Type.method")
    ap.add_argument("--out", help="output JSON")
    ap.add_argument("--index", help="write line->name index here instead")
    args = ap.parse_args()

    src = Path(args.src).resolve()
    repo = Path(args.repo).resolve()
    if args.index:
        index_mode(src, repo, Path(args.index))
        return

    type_name, method = args.target.rsplit(".", 1)
    files = py_files(src)
    project = jedi.Project(str(repo))
    rel = lambda p: Path(p).resolve().relative_to(repo).as_posix()

    class_defs, subclasses = build_class_index(files)
    if type_name not in class_defs:
        raise SystemExit(f"target class {type_name!r} not found under {src}")

    # Family: declaration + every transitive subclass that defines `method`.
    fam_classes = [type_name] + sorted(transitive_subclasses(type_name, subclasses))
    family_defs = []  # (file, ast def node)
    for cname in fam_classes:
        for f, cls in class_defs.get(cname, []):
            m = method_in_class(cls, method)
            if m is not None:
                family_defs.append((f, m))
    if not family_defs:
        raise SystemExit(f"no class in the {type_name} hierarchy defines {method!r}")

    family_sites, caller_sites = set(), set()
    trees: dict[Path, ast.Module] = {}

    def tree_of(f: Path) -> ast.Module:
        if f not in trees:
            trees[f] = ast.parse(f.read_text(), filename=str(f))
        return trees[f]

    for f, mnode in family_defs:
        family_sites.add(f"{rel(f)}:{method}")

    # Callers: Jedi project-wide references to each family def, in call
    # position, attributed to their enclosing def. Jedi resolves the receiver
    # type, so same-named methods on unrelated classes are excluded.
    for f, mnode in family_defs:
        col = mnode.col_offset + len("async def " if isinstance(mnode, ast.AsyncFunctionDef) else "def ")
        try:
            script = jedi.Script(path=str(f), project=project)
            refs = script.get_references(mnode.lineno, col, include_builtins=False, scope="project")
        except Exception:
            continue
        for r in refs:
            if r.module_path is None:
                continue
            rp = Path(r.module_path)
            rrel = rp.as_posix()
            if "/test" in rrel or rp.name.startswith("test_"):
                continue
            try:
                rp.resolve().relative_to(repo)
            except ValueError:
                continue  # outside repo (stdlib / site-packages)
            # Exclude the definition sites themselves (def name / class body).
            try:
                line_txt = rp.read_text().splitlines()[r.line - 1]
            except (OSError, IndexError):
                continue
            after = line_txt[r.column + len(method):].lstrip()
            if not after.startswith("("):
                continue  # not a call
            if line_txt[:r.column].rstrip().endswith("def"):
                continue  # a definition, not a call
            try:
                site = f"{rel(rp)}:{enclosing_def(tree_of(rp), r.line)}"
            except (OSError, SyntaxError):
                continue
            if site not in family_sites:
                caller_sites.add(site)

    gt = sorted(family_sites | caller_sites)
    result = {
        "target": args.target,
        "family": sorted(family_sites),
        "callers": sorted(caller_sites),
        "ground_truth": gt,
        "stats": {"family": len(family_sites), "resolved_call_hits": len(caller_sites),
                  "gt_sites": len(gt), "family_classes": len(family_defs)},
    }
    Path(args.out).write_text(json.dumps(result, indent=2) + "\n")
    print(json.dumps(result["stats"]))


if __name__ == "__main__":
    main()
